File: 79-word-search/word-search.java
class Solution {
    int[] delrow= {0,-1,0,1};
    int[] delcol= {-1,0,1,0};
    public boolean exist(char[][] board, String word) {
        int n= board.length;
        int m= board[0].length;
        int[][] visited= new int[n][m];

        for(int i=0 ; i< n ;i++){
            for(int j=0 ; j< m ;j++){
                if(word.charAt(0)==board[i][j]){
                     if(dfs(i,j, 0, visited, word,board)){
                    return true;
                    }
                }
               
            }
        }

        return false;

    }
    private boolean dfs(int row, int col,int index, int[][] visited, String word, char[][] board){
        if(word.length()==index) return true;
        if(row<0 || row>= board.length || col<0 || col>= board[0].length || visited[row][col]==1 || board[row][col]!=word.charAt(index) ){
            return false;
        }
        visited[row][col]=1;
        for( int i=0 ; i< 4; i++){
            int nrow= row+delrow[i];
            int ncol= col+delcol[i];

            if(dfs(nrow, ncol, index+1, visited, word, board)){
                return true;
            }
            
        }

        visited[row][col]=0;
        return false;
    }
}

File: 79-word-search/word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        n=len(board)
        m=len(board[0])

        visited=[[0]*m for _ in range(n)]
        directions=[(1,0),(0,1),(-1,0),(0,-1)]

        def dfs(row, col, index, word, board):
            if index == len(word):
                return True
            visited[row][col]=1

            for r, c in directions:
                nrow= r+row
                ncol= c+col
                if(0<= nrow< n and 0<= ncol< m and visited[nrow][ncol]==0 and word[index]==board[nrow][ncol]) :
                    if dfs(nrow, ncol, index+1, word, board):
                        return True
            visited[row][col]=0
            return False
            


        for i in range(n):
            for j in range(m):
                if board[i][j]== word[0]:
                    if dfs(i, j, 1, word, board) :
                        return True
        
        return False
                    


        